### PR TITLE
itemized extrinisc implementation

### PR DESCRIPTION
**Pull request by aramikm**
# Goal
The goal of this PR is to implement an extrinsic for itemized storage changes

Closes issue #925 

# Discussion

# Checklist
- [x] Tests added
- [x] Benchmarks added
- [x] Weights updated

### CODE REVIEW

File: pallets/stateful-storage/src/lib.rs [hunk 1]
```diff
@@ -89,6 +111,18 @@ pub mod pallet {
...
+		#[pallet::constant]
+		type MaxItemizedActionsCount: Get<u32>;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		/// A set of helper functions for benchmarking.
+		type MsaBenchmarkHelper: MsaBenchmarkHelper<Self::AccountId>;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		/// A set of helper functions for benchmarking.
+		type SchemaBenchmarkHelper: SchemaBenchmarkHelper;
```
Comment (saraswatpuneet) [inline pallets/stateful-storage/src/lib.rs:125]: Added MSA and Schema helper for benchmark needs

File: scripts/run_all_benchmarks.sh [hunk 2]
```diff
@@ -6,8 +6,9 @@ PROJECT=${1:-$THIS_DIR/..}
 RUNTIME=$PROJECT/target/production/frequency
 BENCHMARK="$RUNTIME benchmark pallet "
 # TODO: pallet_collator_selection benchmarks fail due to errors in the actual benchmark code. See Issue #608
-EXTERNAL_PALLETS=(pallet_collective orml_vesting pallet_balances pallet_timestamp pallet_session pallet_scheduler pallet_democracy pallet_treasury pallet_preimage pallet_utility)
 CUSTOM_PALLETS=(messages msa schemas stateful-message-storage)
+EXTERNAL_PALLETS=(pallet_collator_selection pallet_collective orml_vesting pallet_balances pallet_timestamp pallet_session pallet_scheduler pallet_democracy pallet_treasury pallet_preimage pallet_utility)
+CUSTOM_PALLETS=(messages msa schemas)
```
Comment (JoeCap08055) [inline scripts/run_all_benchmarks.sh:11]: Duplicate line.

File: pallets/stateful-storage/src/lib.rs [hunk 3]
```diff
@@ -110,35 +144,101 @@ pub mod pallet {
...
 
 	#[pallet::event]
 	#[pallet::generate_deposit(pub(super) fn deposit_event)]
 	pub enum Event<T: Config> {
-		ItemAppended,
-		ItemRemoved,
-		PageAppended,
-		PageUpdated,
-		PageRemoved,
+		PageUpdated { msa_id: MessageSourceId, schema_id: SchemaId },
```
Comment (JoeCap08055) [inline pallets/stateful-storage/src/lib.rs:167]: Maybe add `page_id: PageId[]` to the event? Or not worth it?

Comment (aramikm) [reply pallets/stateful-storage/src/lib.rs:167]: I think we will definitely want page number for paginated ones. maybe an `Optional<PageNumber>`

File: pallets/stateful-storage/src/lib.rs [hunk 4]
```diff
@@ -110,35 +144,101 @@ pub mod pallet {
...
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
 		#[pallet::weight(Weight::default())]
-		pub fn add_item(_origin: OriginFor<T>, payload: Vec<u8>) -> DispatchResult {
+		pub fn apply_item_actions(
+			origin: OriginFor<T>,
+			state_owner_msa_id: MessageSourceId,
+			#[pallet::compact] schema_id: SchemaId,
+			actions: BoundedVec<ItemAction, T::MaxItemizedActionsCount>,
```
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply pallets/stateful-storage/src/lib.rs:179]: i agree with that, earlier I was think to do a mix of actions, but I believe lets follow single responsibility

File: pallets/stateful-storage/src/child_tree_storage.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
+use codec::{Decode, Encode};
```
Comment (JoeCap08055) [inline pallets/stateful-storage/src/child_tree_storage.rs:1]: This file is a duplicate... check `private-graph-main` branch; it's `stateful_child_tree.rs`

File: pallets/stateful-storage/src/lib.rs [hunk 6]
```diff
@@ -49,12 +49,17 @@ mod tests;
...
 mod benchmarking;
+#[cfg(feature = "runtime-benchmarks")]
+use common_primitives::benchmarks::{MsaBenchmarkHelper, SchemaBenchmarkHelper};
 
 mod stateful_child_tree;
 pub mod types;
 
 pub mod weights;
 
+pub mod child_tree_storage;
```
Comment (JoeCap08055) [inline pallets/stateful-storage/src/lib.rs:60]: Change to `stateful_child_tree`

File: runtime/common/src/constants.rs [hunk 7]
```diff
@@ -304,13 +304,21 @@ parameter_types! {
...
+	/// The maximum size of a page (in bytes) for an Itemized storage model
+	pub const MaxItemizedPageSizeBytes: u32 = 4096;
+	/// The maximum size of a page (in bytes) for a Paginated storage model
+	pub const MaxPaginatedPageSizeBytes: u32 = 4096;
+	/// The maximum size of a single item in an itemized storage model (in bytes)
+	pub const MaxItemizedBlobSizeBytes: u32 = 512;
+	/// The maximum number of pages in a Paginated storage model
+	pub const MaxPaginatedPageCount: u32 = 64;
+	/// The maximum number of actions in itemized actions
+	pub const MaxItemizedActionsCount: u32 = 7;
```
Comment (JoeCap08055) [inline runtime/common/src/constants.rs:316]: I guess at some point we should make a more informed determination of what these limits should be?

Comment (aramikm) [reply runtime/common/src/constants.rs:316]: Correct! I think weights 2.0 and PoV size would have a direct effect on these numbers. So we have to make that decision after those are out.

**Review by JoeCap08055 — APPROVED** [review]
Approving to unblock so we can iterate.

**Pull request merged**